include/object/tcb.h

```diff
@@ -128,42 +128,68 @@ void remoteTCBStall(tcb_t *tcb);
 #define SCHED_APPEND_CURRENT_TCB    tcbSchedAppend(NODE_STATE(ksCurThread))
 
 #ifdef CONFIG_KERNEL_MCS
-/* Add TCB into the priority ordered endpoint queue */
-static inline tcb_queue_t tcbEPAppend(tcb_t *tcb, tcb_queue_t queue)
+
+static inline bool_t PURE higher_than_tcb_prio(tcb_t *tcb, prio_t new_priority)
 {
-    /* start at the back of the queue as FIFO is the common case */
-    tcb_t *before = queue.end;
-    tcb_t *after = NULL;
-
-    /* find a place to put the tcb */
-    while (unlikely(before != NULL && tcb->tcbPriority > before->tcbPriority)) {
-        after = before;
-        before = after->tcbEPPrev;
-    }
+    return tcb != NULL && new_priority > tcb->tcbPriority;
+}
 
-    if (unlikely(before == NULL)) {
-        /* insert at head */
-        queue.head = tcb;
-    } else {
-        before->tcbEPNext = tcb;
-    }
+/* Find the rightmost TCB in a queue that has a priority which is strictly
+   greater than the given priority */
+static tcb_t *find_tcb_with_higher_prio(tcb_t *tcb, prio_t new_priority)
+{
+    tcb_t *before = tcb;
 
-    if (likely(after == NULL)) {
-        /* insert at tail */
-        queue.end = tcb;
-    } else {
-        after->tcbEPPrev = tcb;
+    while (higher_than_tcb_prio(before, new_priority)) {
+        before = before->tcbSchedPrev;
     }
 
-    tcb->tcbEPNext = after;
-    tcb->tcbEPPrev = before;
+    return before;
+}
 
-    return queue;
+/* Insert a TCB into a queue immediately after another item in the queue
+   (the queue must initially contain at least two items) */
+static inline void tcb_queue_insert_after(tcb_t *tcb, tcb_t *before)
+{
+    tcb_t *after;
+    after = before->tcbSchedNext;
+
+    assert(after != NULL);
+    assert(after != before);
+
+    tcb->tcbSchedPrev = before;
+    tcb->tcbSchedNext = after;
+
+    after->tcbSchedPrev = tcb;
+    before->tcbSchedNext = tcb;
 }
 
-tcb_queue_t tcbEPDequeue(tcb_t *tcb, tcb_queue_t queue);
+/* Add TCB into the priority ordered endpoint or notification queue */
+static inline tcb_queue_t tcbIPCAppend(tcb_t *tcb, tcb_queue_t queue)
+{
+    prio_t new_priority;
+    tcb_queue_t new_queue;
+
+    new_priority = tcb->tcbPriority;
+    new_queue = queue;
+
+    if (tcb_queue_empty(queue) || new_priority > queue.head->tcbPriority) {
+        new_queue = tcb_queue_prepend(queue, tcb);
+    } else {
+        if (queue.end->tcbPriority >= new_priority) {
+            new_queue = tcb_queue_append(queue, tcb);
+        } else {
+            tcb_t *before;
+            before = find_tcb_with_higher_prio(queue.end, new_priority);
+            tcb_queue_insert_after(tcb, before);
+        }
+    }
+
+    return new_queue;
+}
```

Comment:
```suggestion
static inline tcb_queue_t tcbIPCAppend(tcb_queue_t queue, tcb_t *tcb)
{
    prio_t priority = tcb->tcbPriority;

    if (higher_than_tcb_prio(queue.head, priority)) {
        queue = tcb_queue_prepend(queue, tcb);
    } else {
        if (queue.end->tcbPriority >= priority) {
            queue = tcb_queue_append(queue, tcb);
        } else {
            tcb_t *before = find_tcb_with_higher_prio(queue.end, priority);
            tcb_queue_insert_after(tcb, before);
        }
    }
    return queue;
}
```
The end check could be replaced with `!higher_than_tcb_prio(queue.end, priority)`, but that would add an extra NULL check. Anyway, not sure if using higher_than_tcb_prio() is better, but if not here, then why have it at all?